clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala

```diff
@@ -95,42 +109,63 @@ object UncommittedGarbageCollector {
                         UncommittedGarbageCollector.UNCOMMITTED_GC_SOURCE_NAME
                        )
     val apiClient = ApiClient.get(apiConf)
+    val storageType = apiClient.getBlockstoreType()
     var storageNamespace = apiClient.getStorageNamespace(repo, StorageClientType.HadoopFS)
     if (!storageNamespace.endsWith("/")) {
       storageNamespace += "/"
     }
 
     try {
-      // Read objects directly from object storage
-      val dataDF = listObjects(storageNamespace, cutoffTime)
-
-      // Process uncommitted
-      val uncommittedGCRunInfo =
-        new APIUncommittedAddressLister(apiClient).listUncommittedAddresses(spark, repo)
-
-      var uncommittedDF =
-        if (uncommittedGCRunInfo.uncommittedLocation != "") {
-          spark.read.parquet(uncommittedGCRunInfo.uncommittedLocation)
+      if (shouldMark) {
+        // Read objects directly from object storage
+        val dataDF = listObjects(storageNamespace, cutoffTime)
+
+        // Process uncommitted
+        val uncommittedGCRunInfo =
+          new APIUncommittedAddressLister(apiClient).listUncommittedAddresses(spark, repo)
+
+        var uncommittedDF =
+          if (uncommittedGCRunInfo.uncommittedLocation != "") {
+            spark.read.parquet(uncommittedGCRunInfo.uncommittedLocation)
+          } else {
+            // in case of no uncommitted entries
+            spark.emptyDataFrame.withColumn("physical_address", lit(""))
+          }
+
+        uncommittedDF = uncommittedDF.select(uncommittedDF("physical_address").as("address"))
+        runID = uncommittedGCRunInfo.runID
+
+        // Process committed
+        val committedDF = new NaiveCommittedAddressLister()
+          .listCommittedAddresses(spark, storageNamespace)
+
+        addressesToDelete = dataDF
+          .except(committedDF)
+          .except(uncommittedDF)
+
+        // TODO (niro): not working - need to find the most efficient way to save the first slice
+        if (!dataDF.isEmpty) {
+          val firstFile = dataDF.select(col("address")).first().toString()
+          firstSlice = firstFile.substring(0, firstFile.lastIndexOf("/"))
+        }
+      }
+      removed = {
+        if (shouldSweep) {
+          if (markID != "") { // get the expired addresses from the mark id run
+            addressesToDelete = readMarkedAddresses(storageNamespace, markID)
+          }
+
+          val region = if (args.length == 2) args(1) else null
+          var storageNSForSdkClient = apiClient.getStorageNamespace(repo, StorageClientType.SDKClient)
+          if (!storageNSForSdkClient.endsWith("/")) {
+            storageNSForSdkClient += "/"
+          }
+          val hcValues = spark.sparkContext.broadcast(HadoopUtils.getHadoopConfigurationValues(hc, "fs.", "lakefs."))
+          val configMapper = new ConfigMapper(hcValues)
```

Comment:
Is this logic similar to committed GC? can we extract it to somewhere common?

Reply:
done